Arch arm linker style and arch fixes

ioannisg opened the pull request:
1. Adding a Kconfig symbol to denote the minimum alignment and region size for MPU
2. Using the symbol in ARM arch headers and linker scripts
3. Fixing a hard-coded value in thread stack size truncation

Follow-up PR: cleaning app_data_alignment.ld

ioannisg committed:
arch: arm: Kconfig option for minimum MPU region size and alignment

The commit introduces a Kconfig option to define the minimum MPU
region size and alignment for ARM Cortex-M MCU architecture. The
hidden option is used by the linker to properly align linker
sections respecting the MPU requirements. Additionally, it is
used to define MPU Guard size and minimum stack alignment for
ARM Cortex-M MCUs in include/arch/arm/arch.h .

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: utilize Kconfig option for min ARM MPU region size

The commit enforces the use of ARM_MPU_REGION_MIN_ALIGN_AND_SIZE
in include/arch/arm/arch.h, instead of using 32 as a hard-coded
value. The symbol is also used in arm/thread.c to truncate the
thread stack size to satisfy MPU granularity. The commit does
not introduce behavioral changes.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

AdithyaBaglody commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -283,7 +283,11 @@ SECTIONS
-#include <app_data_alignment.ld>
+#include <linker_section_alignment.ld>
+
+		. = ALIGN(4);
```
I think this can be removed because it will be handled by line 290.

SebastianBoe replied:
Sure? What if _region_min_align is 0?

ioannisg replied:
Right, @SebastianBoe , that's why I kept the (4) here. 

However, it is rather unlikely that _region_min_align will be 0 in future MPU architectures.

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -1078,6 +1078,10 @@ zephyr_compile_options(
 )
 
+configure_file(
+     $ENV{ZEPHYR_BASE}/include/arch/arm/cortex_m/scripts/linker_section_alignment.ld
```
${ZEPHYR_BASE}

ioannisg replied:
Yeap, I can update this. (copy-pasted from following configure_file( statements)

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -1078,6 +1078,10 @@ zephyr_compile_options(
 )
 
+configure_file(
+     $ENV{ZEPHYR_BASE}/include/arch/arm/cortex_m/scripts/linker_section_alignment.ld
```
You don't need a configure_file.

At least I don't see anything in the linker script that requires it.

ioannisg replied:
done as discussed offline; added the linker symbol directly in linker.ld, removed the introduced file and the changes in CMakeLists.txt

SebastianBoe reviewed (changes requested):
minor fixes needed.

ioannisg commented:
Added a commit to clean up the unused symbols, _image_ram_start, _image_ram_end from ARM linker script as suggested by @SebastianBoe

ioannisg commented:
@aurel32 FYI

galak commented:
Would like @agross-linaro to have a chance to review, he's on vacation this week.

agross-linaro commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -283,7 +293,8 @@ SECTIONS
 	}
 #endif
 
-#include <app_data_alignment.ld>
```
I think this needs to remain.  The alignment required here will depend on the region size requirements.  For power of two, that needs to be a power of two align to make sure we don't let anything else get exposed.  During compile, this include file may or may not be generated and may or may not differ from 32.  And we can't use a Kconfig because in some cases we'll be using the app data size used rounded up to the next power of two.

ioannisg replied:
@agross-linaro , yes, I realized that, as well. This needs to be brought back.

And I think the very same is needed for every ram section we add in the linker..

ioannisg replied:
What bothers me, also, is the fact that app_data_alignment is attempting to align the start of next section, assuming that the next section is the Application Memory. But is looks to me that it might, instead, me the APP Shared Memory. I think the inclusion needs to be brought exactly before that section is defined.

aurel32 replied:
I guess we also want to get the python script and the corresponding cmake code to be removed at the same time.

agross-oss reviewed (changes requested):
I'm not convinced this is done correctly.  We cannot use a Kconfig option to define this.  In some cases it is computed from the application data size that is determined by the first pass compile.  We use a script to generate the linked include file which may cause the images to slide a little.

ioannisg committed:
arch: arm: properly enforce MPU-aware linker section alignment

This commit standardizes and simplifies the way we enforce
linker section alignment, to comply with minimum alignment
requirement for MPU, if we build Zephyr with MPU support:
- it enforces alignment with the minimum MPU granularity at
  the beginning and end of linker sections that require to
  be protected by MPU,
- it enforces alignment with size if required by the MPU
  architecture.

Particularly for the Application Memory section, the commit
simplifies how the proper alignment is enforced, removing
the need of calculating the alignment with a post-linker
python script. It also removes the need for an additional
section for padding.

For the Application Shared Memory section(s), the commit
enforces minimum alignment besides the requirement for
alignment with size (for the respective MPUs) and fixes
a bug where the app_data_align was erronously used in the
scipts for auto-generating the linker scripts.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: make _image_ram_start point to the beginning of sram

Move the definition of _image_ram_start at the beginning
of the RAMMABLE (SRAM) region, so it points to the actual
start of RAM linker sections.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -263,6 +273,11 @@ SECTIONS
+	 * minimum granularity required by MPU.
+	 */
+	. = ALIGN(_region_min_align);
+	_image_ram_start = .;
```
_image_ram_start is, now, placed in the right point

SebastianBoe replied:
I thought we were going to get rid of it.

What is it for?

ioannisg replied:
I am going to need this very soon, in ARMv8-M to mark the area from the beginning of SRAM until the first MPU region that corresponds to some kernel feature (e.g. Application Memory). Looks like there might be some SOC-specific linker sections there.

ioannisg commented on scripts/gen_app_partitions.py:
```diff
@@ -12,24 +12,24 @@
 		. = ALIGN( 1 << LOG2CEIL(data_smem_{0}b_end - data_smem_{0}));
 		data_smem_{0} = .;
 		KEEP(*(SORT(data_smem_{0}*)))
-		. = ALIGN(_app_data_align);
```
this looks like a bug: app_data section alignment does not have anything to do with shared app mem, IMHO

AdithyaBaglody replied:
I added this because the next section would need to app_data_aligned. not for this section actually. 
But ya i agree this should be moved out of this script.

ioannisg commented:
> I'm not convinced this is done correctly. We cannot use a Kconfig option to define this. In some cases it is computed from the application data size that is determined by the first pass compile. We use a script to generate the linked include file which may cause the images to slide a little.

@agross-linaro thanks for your review on this, it was really helpful

So, I have now tried to simplify the way we align the Application Memory section: no script is now needed.

Additionally, I have fixed some bugs in the generation of the Shared memory section(s) and placed the `_image_ram_start `in the right place.

I will need @SebastianBoe and @agross-linaro to re-review this one. Might be nice if @andrewboie and @aurel32 could look at it, too.

AdithyaBaglody commented on scripts/gen_app_partitions.py:
```diff
@@ -12,24 +12,24 @@
 linker_end_seq = """
-		_app_smem_end = .;
 		APP_SHARED_ALIGN;
+		_app_smem_end = .;
```
this needs to go back to its original place. 
We cant risk adding extra alignment. There is a possibility that this can screw up region size calculation.

ioannisg replied:
I am afraid I can't understand why we can screw-up here. There won't be extra alignment.
But I suppose i can revert this - it looks equivalent to me, as long as we do not put any symbols after the end of the last app smem section and before _app_smem_end

ioannisg commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -339,15 +334,43 @@ SECTIONS
 	__app_ram_size = __app_ram_end - __app_ram_start;
 #endif /* CONFIG_APPLICATION_MEMORY */
 
+#include <app_data_alignment.ld>
```
Ups, this is forgotten here, yes. I will remove, thanks.

So, yes, in general this script is not needed, now, when we build _for ARM!_
I think it is still used for other platforms.

ioannisg commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -339,15 +334,43 @@ SECTIONS
 
+#include <app_data_alignment.ld>
+
+#if defined(CONFIG_APP_SHARED_MEM)
```
Yes, there is. I want to move _all_ fixed regions in the beginning of SRAM.
(Fixed regions = regions over which we do not do dynamic MPU programming afterwards)
Currently this concerns Application Memory, but soon we might get Region to Execute from RAM and No-Cache region (see relevant PRs) :)

ioannisg replied:
To elaborate a bit more; all linker sections corresponding to "feature" regions, such as Application Memory are moved to the beginning of image RAM, so, Kernel and Shared App memory can be placed in a single, continuous area in SRAM. The latter is essential for ARMv8-M, so that we may have all dynamic MPU regions (such as Stack Guards, thread stacks, memory partitions) be programmed within a single area; this is going to simplify the design.

ioannisg commented on scripts/gen_app_partitions.py:
```diff
@@ -12,24 +12,24 @@
 # when APP_SHARED_MEM is enabled.
 print_template = """
 		/* Auto generated code do not modify */
+		APP_SHARED_ALIGN;
```
I m thinking; what if the next size aligns with a too-small power-of-two? Don't we need to ensure at least 32?

ioannisg commented:
@AdithyaBaglody thanks for re-reviewing.

ioannisg committed:
scripts: add missing lincense in scripts/gen_app_partitions.py

Add missing lincense header in scripts/gen_app_partitions.py

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

aurel32 commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -68,6 +68,16 @@
+ */
+#if defined(CONFIG_ARM_MPU_REGION_MIN_ALIGN_AND_SIZE) \
+	&& (CONFIG_ARM_MPU_REGION_MIN_ALIGN_AND_SIZE != 0)
+_region_min_align = CONFIG_ARM_MPU_REGION_MIN_ALIGN_AND_SIZE;
```
Instead of checking of the config value is not 0, wouldn't it be better to define it directly as 4 so that the same alignment value is used in both the linker files and the code?

aurel32 replied:
For MPUs which require 0-byte alignment, we will still need a a minimum 4-byte alignment both in the linker script and the MPU code. So my point would be to drop the `&& (CONFIG_ARM_MPU_REGION_MIN_ALIGN_AND_SIZE != 0)` and to make the default value of `CONFIG_ARM_MPU_REGION_MIN_ALIGN_AND_SIZE` to 4 instead of 0 on non-ARM MPUs. That way both the linker and the MPU code would use the same value.

ioannisg replied:
Right. But still someone can mess up and overwrite that with 0. So we would need this extra check, imho.

aurel32 replied:
This is actually quite important to have `CONFIG_ARM_MPU_REGION_MIN_ALIGN_AND_SIZE` != 0 in the MPU code otherwise code like this one would simply not work and truncate the size to 0:

```
_current->stack_info.size &=
		~(CONFIG_ARM_MPU_REGION_MIN_ALIGN_AND_SIZE - 1);
```

ioannisg replied:
So, this code block is compiled-in only if MPU support is enabled. So my intention was that, if anyone uses the MPU, and overwrites the definition of CONFIG_ARM_MPU_REGION_MIN_ALIGN_AND_SIZE, they must know what they are doing. Defining =0 does not make sense (there is no 0-alignment requirement. the minimum could be 1).

aurel32 replied:
As you said this is an _hidden_ Kconfig option, so no-one should not be able to mess up. I think it's important to have a consistent value between the MPU code and the linker code. That's actually why it's cleaner to have the value defined in a single place.

ioannisg replied:
Though I agree with your argument, I have a single concern: If anyone does not want to use the MPU (builds without MPU support), why would we need to have CONFIG_ARM_MPU_REGION_MIN_ALIGN_AND_SIZE != 0? I don't see this necessary _in C code_
In the linker, however, we still need alignment-4 for reasons _other than MPU_ That's why the linker enforces 4-byte minimum.

ioannisg replied:
OH, I see. Yes, I am convinced that this can be removed. I agree that if anyone sets CONFIG_ARM_MPU_REGION_MIN_ALIGN_AND_SIZE to zero, deliberately, he should face the consequences of their action.

ioannisg replied:
Fixing this @aurel32

aurel32 commented on arch/arm/core/cortex_m/mpu/Kconfig:
```diff
@@ -17,6 +17,14 @@ config ARM_MPU
 	  Note that NXP MPU as well as ARMv8-M MPU does not require MPU regions
 	  to have power-of-two alignment for base address and region size.
 
+config ARM_MPU_REGION_MIN_ALIGN_AND_SIZE
```
It's nice to remove the hardcoded value from the code, but I wonder if Kconfig is the right place for this value. Wouldn't it be better if this is put in one of the _internal.h files instead?

ioannisg replied:
This was discussed with @SebastianBoe . The reason for a Kconfig _hidden_ symbol, was to make both the C code and the linker scripts use a common reference for this fundamental MPU property, instead of having linker scripts including C headers. This was seen as a cleaner option.

aurel32 replied:
Ok

aurel32 commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -339,15 +334,41 @@ SECTIONS
 #endif /* CONFIG_APPLICATION_MEMORY */
 
+#if defined(CONFIG_APP_SHARED_MEM)
+#if defined(CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT)
```
I don't get why we still need to get two different strategies for application shared memory depending if the MPU requires power of two alignement or not.

ioannisg replied:
@AdithyaBaglody could you advice here?
We could possibly try to use _only_ the script for both cases of MPU architectures.

(The difference with APPLICATION_MEMORY is that, in this case, we have multiple sections.)

aurel32 reviewed (commented):
In order to avoid repeated code, I wonder if we could define a MPU_ALIGN macro at the beginning of the script which could be something like that (untested):

```
#if defined(CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT)
#define MPU_ALIGN(region_size) \
    . = ALIGN(_region_min_align);
    . = ALIGN( 1 << LOG2CEIL(region_size);
#else
#define MPU_ALIGN(region_size) \
    . = ALIGN(_region_min_align);
#endif
```
Then it's just a question of adding a single MPU_ALIGN(__section_end - __section_start) to handle all cases.

ioannisg commented:
>In order to avoid repeated code, I wonder if we could define a MPU_ALIGN macro at the beginning of the script which could be something like that (untested):
#if defined(CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT)
#define MPU_ALIGN(region_size) \
    . = ALIGN(_region_min_align);
    . = ALIGN( 1 << LOG2CEIL(region_size);
#else
#define MPU_ALIGN(region_size) \
    . = ALIGN(_region_min_align);
#endif
Then it's just a question of adding a single MPU_ALIGN(__section_end - __section_start) to handle all cases.

@aurel32 that's a good suggestion. I will implement this.

ioannisg commented:
@aurel32 I implemented your suggestions. Pls, take one more look when possible.

ioannisg commented:
@AdithyaBaglody @agross-linaro could you revisit your reviews?

AdithyaBaglody reviewed (approved):
LGTM

ioannisg commented on arch/arm/core/thread.c:
```diff
@@ -122,11 +122,9 @@ FUNC_NORETURN void _arch_user_mode_enter(k_thread_entry_t user_entry,
-	_current->stack_info.size &= ~0x1f;
+	/* Truncate the stack size with the MPU region granularity. */
+	_current->stack_info.size &=
+		~(CONFIG_ARM_MPU_REGION_MIN_ALIGN_AND_SIZE - 1);
```
@andrewboie FYI (as author of this "FIXME")

aurel32 commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -283,34 +307,11 @@ SECTIONS
 	}
 #endif
 
-#include <app_data_alignment.ld>
```
With that change done, I guess you should also:
- git rm include/arch/arm/cortex_m/scripts/app_data_alignment.ld
- remove the 3 lines: https://github.com/zephyrproject-rtos/zephyr/blob/master/CMakeLists.txt#L1080
- change the condition https://github.com/zephyrproject-rtos/zephyr/blob/master/CMakeLists.txt#L1113 to
`if(CONFIG_ARC and CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT AND CONFIG_APPLICATION_MEMORY)`

ioannisg replied:
> git rm include/arch/arm/cortex_m/scripts/app_data_alignment.ld

Is it guaranteed that the script is not used by other architectures, e.g. ARC?

aurel32 replied:
This file is in an ARM specific directory, and I really doubt that the ARC include paths contains an ARM directory.

aurel32 reviewed (commented):
Thanks for the updated version and for taking the comments into account. I only have 2 minor comments left, it looks almost ready to me.

ioannisg commented on arch/arm/core/cortex_m/mpu/Kconfig:
```diff
@@ -17,6 +17,14 @@ config ARM_MPU
+config ARM_MPU_REGION_MIN_ALIGN_AND_SIZE
+	int
+	default 32 if ARM_MPU
+	default 0
```
Right, I can do that.

ioannisg committed:
cmake: move app_data_alignment.ld under ARC sub-directory

This commit moves the app_data_alignment.ld scripts
under arch/arc sub-directory, as it is not not used
at all in ARM builds. The script is still used for
ARC, whose v2 MPU also has the reuquirement for
power-of-two size alignment.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
@ruuddw , please, review the last commit of this PR, which moves the app_data_alignment.ld script under ARC (as it is not not used in ARM builds)

@aurel32 @agross-linaro please, have a final look into this!

ioannisg committed:
cmake: do not run app data section aligning post-script for ARM

Since the app_data_alignment.ld is not used anymore in ARM
builds, remove the need for running the script to auto-
generate it. The script now runs only for ARC builds.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

tyrantlizard77 reviewed (approved):
LGTM

agross-oss reviewed (approved):
lgtm.  had to run a few passes to see how it worked out.
